Bootstrap and wipe granularity and #538 Fix

RobertSzkutak opened the pull request:
The purpose of this pull request is to allow users to specify which parts of their applications to bootstrap using the bootstrap command.

Examples: 

./ml local bootstrap --apply-changes=indexes,users,roles
(will only bootstrap the indexes, users, and roles)

./ml local wipe --apply-changes=databases,forests,appservers
(will only wipe databases, forests, and app servers)

This pull request accomplishes several things to make this possible:

* Index configuration is now it's own function in setup.xqy . It was split out of database configuration.
* setup:do-setup changed to respect the additional bootstrap options
* server_config.rb bootstrap changed to handle the --apply-changes option and pass it on to setup:do-setup

I haven't tested everything and would appreciate a second set of eyes to test as well as feedback on any ways that the code could be improved.

=============

Additionally, this branch fixes this bug with CORB: https://github.com/marklogic/roxy/issues/538

grtjn commented on deploy/lib/server_config.rb:
```diff
@@ -949,7 +955,7 @@ def load_data(dir, options = {})
 
     options[:batch_commit] = batch
-    options[:permissions] = permissions(@properties['ml.app-role'], Roxy::ContentCapability::ER) unless options[:permissions]
+    options[:permissions] = permissions(@properties['ml.app-role'], Roxy::ContentCapability::ERU) unless options[:permissions]
```
Looks like there are a few redundant changes in this PR. Did you start off with upstream/dev, and did you fetch and rebase against that as well?

RobertSzkutak replied:
Yes, I did an upstream fetch from dev and then rebased against that. After that I began to make changes. Im dont think my other pull request for dev with this fix was accepted which may be why it was still lingering in my tree?

grtjn commented on deploy/default.properties:
```diff
@@ -1,222 +1,223 @@
+http.retry-count=3
+http.open-timeout=5
+http.read-timeout=300
+http.retry-delay=15
```
Odd, unexpected diff..

RobertSzkutak replied:
Agreed, was surprised to see this too. I have no idea what I did :-)

grtjn commented on deploy/lib/xquery/setup.xqy:
```diff
@@ -469,31 +469,36 @@ declare function setup:rewrite-config($import-configs as element(configuration)+
-      setup:create-appservers($import-config),
-      setup:apply-appservers-settings($import-config),
-      setup:create-scheduled-tasks($import-config),
+      if(fn:exists(fn:index-of($options, "all")) or fn:exists(fn:index-of($options, "ssl"))) then setup:create-ssl-certificate-templates($import-config) else (),
```
The exists(index-of(..)) makes this very difficult to read. How about a little FLWOR on top that converts $options into a hash-map in which you do one simple if (map:exists($options, 'ssl')) then.. (and make the $options such that if $options was empty, all keys are true by default..)

RobertSzkutak replied:
Sounds great! I'll get to work!

grtjn commented:
The change looks relatively simple. Can we do the same trick with wipe? That would useful for testing this too..

RobertSzkutak commented:
Great idea! I'll build that in along with your suggested cleanup for checking the options.

RobertSzkutak commented:
I added in three more commits which make the changes you recommended as well as add identical functionality to wipe.

joemfb commented:
Here's what I usually do to fix up these kinds of rebase issues:

    git checkout dev
    git pull upstream dev
    git checkout rob
    git rebase dev
    git push origin rob -f

If that doesn't correctly isolate your commits, you may need to create a new branch and cherry-pick your commits specifically.

RobertSzkutak commented:
No luck, unfortunately :\

```
$ git checkout dev
Switched to branch 'dev'
$ git pull upstream dev
From https://github.com/marklogic/roxy
 * branch            dev        -> FETCH_HEAD
Already up-to-date.
$ git checkout rob
Switched to branch 'rob'
$ git rebase dev
Current branch rob is up to date.
```

Any other ideas?

I can take a look tonight at carefully constructing a cleaner branch.

joemfb commented:
Yeah, it looks like you'll need to. There's some good answers here: https://stackoverflow.com/questions/1994463/how-to-cherry-pick-a-range-of-commits-and-merge-into-another-branch.

I favor the three argument form of `git rebase --onto`; there's some good details here: https://blog.pivotal.io/labs/labs/git-rebase-onto.

grtjn commented:
I sometimes get unexpected commits appearing after squashing commits with git rebase -i. I usually get rid of them by doing another git rebase (without -i)

RobertSzkutak commented:
All fixed now?

I made one additional change to fix a merge conflict with this branch and the branch pulled in yesterday (the #514 fix was not included in the branch pulled in yesterday)

joemfb commented:
I'm still seeing some older commits (like 4a276a from rloupre)...

 You want to create a new branch from the latest `marklogic/dev` branch, and then populate with only the commits for this PR. You could accomplish that be editing your `rob` branch in place, or maybe just by creating a new branch from `marklogic/dev` and individually rebasing/cherry-picking your relevant commits from `rob`.

RobertSzkutak commented:
Third time is the charm?

Thank you for patience with me. I really appreciate your help.

RobertSzkutak commented:
Bump. Any chance of merging this in soon? One of our largest clients is using this pretty heavily right now and have been asking about when/if my changes will be "officially" supported.

dmcassel commented:
Looks good to me. @grtjn have you already tested this?

joemfb commented:
FYI, the bulk of the changes in `default.properties` are line endings (you can ignore whitespace in diffs by adding `?w=0` to this PR URL).

dmcassel commented:
Just ran a test. 

    $ ./ml local bootstrap
    $ ./ml local wipe --apply-changes=roles

This successfully removes `roxy-role`, without removing `roxy-user`. 

    $ ./ml local bootstrap --apply-changes=roles

This restores `roxy-role`, but does not re-associate `roxy-user` with it. I think those two commands back-to-back should result in the same state as a full bootstrap.

dmcassel commented:
Should this work? 

    $ ./ml local wipe --apply-changes=forests

I get 

> ERROR: ADMIN-DATABASEFORESTATTACHED: (err:FOER0000) Forest 17077807573084922939 is attached to a database

I think "wipe forests" should detach and remove and "bootstrap forests" should create and attach.

dmcassel commented:
$ ./ml bootstrap -h

and

    $ ./ml wipe -h 

should show the --apply-changes option, complete with what options are available. That's a pretty simple change in deploy/Help.rb.

dmcassel commented:
@RobertSzkutak, thanks for persisting with this, looks like a handy capability.

RobertSzkutak commented:
Thanks for testing! We've been mostly focused on roles, users, and index usage at my client and I hadn't spotted these yet. I'll continue to improve this branch and have some new commits soon.

RobertSzkutak commented:
Merged in a pull request to my branch from Josh Meekhof which fixes the user and role assignment issue pointed out by @dmcassel .

RobertSzkutak commented:
Added help documentation courtesy of @jmeekhof

RobertSzkutak commented:
Pushed in a fix for this issue: https://github.com/marklogic/roxy/issues/538

My client has been using this branch very heavily and needed the fix immediately.

dmcassel commented:
This passes my tests now. Thanks for submitting, @RobertSzkutak
